common-healing.lic

```diff
@@ -2,1089 +2,135 @@
+    when 'feel only an aching emptiness'
+      # Empath is permashocked and cannot perceive wounds.
+      return
     end
 
-    { 'wounds' => wounds, 'poisoned' => poisoned, 'diseased' => diseased, 'parasites' => parasites }
-  end
+    pause 0.5
+    health_lines = reget(100).map(&:strip).reverse
+
+    # Example of what we're parsing:
+    # ===
+    # Your injuries include...
+    # Wounds to the LEFT ARM:
+    # Fresh External:  cuts and bruises about the left arm -- more than minor
+    # Scars External:  severe scarring and chunks of flesh missing from the left arm -- very severe
+    # Fresh Internal:  minor swelling and bruising around the left arm -- more than minor
+    # Scars Internal:  a completely paralyzed left arm -- useless
+    # ===
+    if health_lines.grep(/^Your injuries include\b/).any?
+      perceived_wounds = Hash.new { |h, k| h[k] = [] }
+      body_part = nil
+      health_lines
+        .take_while { |line| line != 'Your injuries include...' }
+        .reverse # put back in normal order so we parse <body part> line then <wound> lines
+        .each do |line|
+          case line
+          when /^Wounds to the (.+):/
+            body_part = Regexp.last_match(1)
+            perceived_wounds[body_part] = []
+          when /^(Fresh|Scars) (External|Internal)/
+            # Do regex then grab matches for wound details.
+            line =~ $DRCH_PERCEIVE_HEALTH_SEVERITY_REGEX
+            severity = $DRCH_WOUND_TO_SEVERITY_MAP[Regexp.last_match[:severity]]
+            is_internal = Regexp.last_match[:location] == 'Internal'
+            is_scar = Regexp.last_match[:freshness] == 'Scars'
+            perceived_wounds[body_part] << Wound.new(
+              body_part: body_part,
+              severity: severity,
+              is_internal: is_internal,
+              is_scar: is_scar
+            )
+          end
+        end
+    end
 
-  def check_perc_health
-    return unless DRStats.empath?
-    DRC.bput('perc heal self', 'Roundtime')
-    data = reget 100
-    data = data.reverse.take_while { |item| item != "Your injuries include...\r" }.reverse
-    wounds = {}
-    part = nil
-    data.each do |line|
-      if line =~ /Wounds to the (.+):/
-        part = Regexp.last_match(1)
-        wounds[part] = []
-      end
-      wounds[part] += [{ 'part' => part, 'location' => Regexp.last_match(2), 'type' => Regexp.last_match(1), 'level' => $wound_map[Regexp.last_match(3)] }] if line =~ /(\w+) (\w+):.*\-\- (.*)\r/
+    # Now go through the preceived wounds map (key=body part)
+    # and bucket the wounds by severity into the 'wounds' map (key=severity).
+    perceived_wounds.values.flatten.each do |wound|
+      wounds[wound.severity] << wound
     end
+
+    # The perceive command incurs roundtime.
+    # We delayed waiting until now so that
+    # the script can use that time to parse
+    # the output rather than waiting unnecessarily.
     waitrt?
-    wounds
+
+    return {
+      'wounds' => wounds
```

Comment:
Might be nice to do the non-rt inducing check_health, and present the empath back the same structure, just more detailed for wounds.

Reply:
I'm not sure I follow, can you expand on this idea?

Reply:
It might be nice for empath scripts to not have to call both and squish them together. This function could call the normal check_health, and then replace the check_health.wounds with the new (more accurate) perc_health.wounds, and return the full wound state back to the empath; i.e. 

{
 'wounds' => wounds,
 'bleeders' => check_health_response.bleeders,
 'parasites' => check_health_response.parasites,
 etc.
}

Reply:
Ahh, I understand now, thanks. I'll do that